node/network/bitfield-distribution/src/lib.rs

```diff
@@ -0,0 +1,684 @@
+const COST_MESSAGE_NOT_DECODABLE: ReputationChange =
+	ReputationChange::new(-100, "Not intersted in that parent hash");
+
+/// Checked signed availability bitfield that is distributed
+/// to other peers.
+#[derive(Encode, Decode, Debug, Clone, PartialEq, Eq)]
+pub struct BitfieldGossipMessage {
+	/// The relay parent this message is relative to.
+	pub relay_parent: Hash,
+	/// The actual signed availability bitfield.
+	pub signed_availability: SignedAvailabilityBitfield,
+}
+
+/// Data used to track information of peers and relay parents the
+/// overseer ordered us to work on.
+#[derive(Default, Clone)]
+struct Tracker {
+	/// track all active peers and their views
+	/// to determine what is relevant to them.
+	peer_views: HashMap<PeerId, View>,
+
+	/// Our current view.
+	view: View,
+
+	/// Additional data particular to a relay parent.
+	per_relay_parent: HashMap<Hash, PerRelayParentData>,
+}
+
+/// Data for a particular relay parent.
+#[derive(Debug, Clone, Default)]
+struct PerRelayParentData {
+	/// Signing context for a particular relay parent.
+	signing_context: SigningContext<Hash>,
+
+	/// Set of validators for a particular relay parent.
+	validator_set: Vec<ValidatorId>,
+
+	/// Set of validators for a particular relay parent for which we
+	/// received a valid `BitfieldGossipMessage`.
+	/// Also serves as the list of known messages for peers connecting
+	/// after bitfield gossips were already received.
+	one_per_validator: HashMap<ValidatorId, BitfieldGossipMessage>,
+
+	/// which messages of which validators were already sent
+	message_sent_to_peer: HashMap<PeerId, HashSet<ValidatorId>>,
+}
+
+impl PerRelayParentData {
+	/// Determines if that particular message signed by a validator is needed by the given peer.
+	fn message_from_validator_needed_by_peer(
+		&self,
+		peer: &PeerId,
+		validator: &ValidatorId,
+	) -> bool {
+		if let Some(set) = self.message_sent_to_peer.get(peer) {
+			!set.contains(validator)
+		} else {
+			false
+		}
+	}
+}
+
+fn network_update_message(n: NetworkBridgeEvent) -> AllMessages {
+	AllMessages::BitfieldDistribution(BitfieldDistributionMessage::NetworkBridgeUpdate(n))
+}
+
+/// The bitfield distribution subsystem.
+pub struct BitfieldDistribution;
+
+impl BitfieldDistribution {
+	/// The protocol identifier for bitfield distribution.
+	const PROTOCOL_ID: ProtocolId = *b"bitd";
+
+	/// Start processing work as passed on from the Overseer.
+	async fn run<Context>(mut ctx: Context) -> SubsystemResult<()>
+	where
+		Context: SubsystemContext<Message = BitfieldDistributionMessage>,
+	{
+		// startup: register the network protocol with the bridge.
+		ctx.send_message(AllMessages::NetworkBridge(
+			NetworkBridgeMessage::RegisterEventProducer(Self::PROTOCOL_ID, network_update_message),
+		))
+		.await?;
+
+		// work: process incoming messages from the overseer and process accordingly.
+		let mut tracker = Tracker::default();
+		loop {
+			{
+				let message = ctx.recv().await?;
+				match message {
```

Comment:
On code structure: branches here get indented a lot and probably go beyond 100 characters. Beyond that, there would be other benefits to breaking out the logic for things here into separate functions. I would find the code easier to read & review, and it would be possible to test individual branches without sending a message in.

So maybe a

```rust
async fn handle_message(..., BitfieldDistributionMessage);
async fn handle_network_update(...);
async fn handle_overseer_signal(...);
```

and then the main `async fn run` just delegates to these.